ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,188 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+
+	"github.com/celestiaorg/celestia-node/ipld/plugin"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// TODO(@distractedm1nd) Find a better figure than NumWorkersLimit for this pool.
+// Worker pool responsible for the goroutines spawned by getLeavesByNamespace
+var namespacePool = workerpool.New(NumWorkersLimit)
+
+// GetSharesByNamespace walks the tree of a given root and returns its shares within the given namespace.ID.
+func GetSharesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]Share, error) {
+	leaves, err := getLeavesByNamespace(ctx, bGetter, root, nID, maxShares)
+	if err != nil {
+		return nil, err
+	}
+
+	shares := make([]Share, len(leaves))
+	for i, leaf := range leaves {
+		shares[i] = leafToShare(leaf)
+	}
+
+	return shares, nil
+}
+
+// wrappedWaitGroup is needed because waitgroups do not expose their internal counter,
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	jobs    chan *job
+	counter int64
+}
+
+func createWrappedWaitGroup(jobs chan *job) *wrappedWaitGroup {
```

Comment:
```suggestion
func newWrappedWaitGroup(jobs chan *job) *wrappedWaitGroup {
```

Reply:
This can be generified and extracted into `libs` as a utility 😃 . I see this being reusable. Up to you to decide whether to extract now, later, or not at all.

Reply:
Hmmm. I feel like the inclusion of the jobs channel makes this not very suitable for re-use. I would wait to extract this until we want to use it somewhere else, since it smells like premature abstraction

Reply:
I am just going to get rid of this method, I imagined it better than it actually is 😅

Reply:
> I feel like the inclusion of the jobs channel makes this not very suitable for re-use.
That's why generify. 

>  I would wait to extract this until we want to use it somewhere else, since it smells like premature abstraction
Yeah. I've written similar code before, which closes the chan once the atomic counter is zero, and this reminds me of it and makes me wanna extract it, but as I said, completely up to you.